ui/src/app/components/admin/audit-page-component.tsx

```diff
@@ -0,0 +1,147 @@
+import {AuditActionCardListView} from 'app/components/admin/audit-card-list-view';
+import {Navigate} from 'app/components/app-router';
+import {Button} from 'app/components/buttons';
+import {NumberInput, TextInputWithLabel} from 'app/components/inputs';
+import colors from 'app/styles/colors';
+import {actionToString} from 'app/utils/audit-utils';
+import {AuditAction, AuditLogEntry} from 'generated';
+import * as fp from 'lodash/fp';
+import * as React from 'react';
+
+const {useEffect, useState} = React;
+
+// The APIs for user and workspace audit are slightly different, and typed differently,
+// because of the need to return different database IDs and not get them confused (in future endpoints).
+// This interface allows an easy mapping from both real query response types.
+export interface GenericAuditQueryResult {
+  /**
+   * Re-constructed objects for each composite action in the query results.
+   */
+  actions?: Array<AuditAction>;
+  logEntries: Array<AuditLogEntry>;
+  /**
+   * Copy of the query used in BigQuery to assist in further exploration or debugging.
+   */
+  query: string;
+  /**
+   * ID in the MySQL database and BigQuery Audit Database for the this qyery. Currently either
+   * a userId or workspaceId as appropriate. This may be needed for situations (such as workspace audit)
+   * where it's not obvious in the responses which workspace is the one you audited.
+   */
+  sourceId: number;
+}
+
+// Common properties for User & Workspace (and similar future pages).
+// Assumptions: the path parameter is called initially with the audit API subject.
+//   (this will need revisiting for multi-subject queries)
+export interface AuditPageProps {
+  initialAuditSubject?: string;
+  auditSubjectType: string;
+  queryAuditLog: (subject: string) => Promise<GenericAuditQueryResult>;
+  getNextAuditPath: (subject: string) => string;
+  logVerbose: boolean;
+  buttonLabel?: string;
+}
+
+const UserInput = ({initialAuditSubject, auditSubjectType, getNextAuditPath, buttonLabel}) => {
+  const [auditSubject, setAuditSubject] = useState(initialAuditSubject);
+  const [loadNextSubject, setLoadNextSubject] = useState(false);
+  useEffect(() =>  {
+    if (loadNextSubject) {
+      setLoadNextSubject(false);
+    }
+  }, [loadNextSubject]);
+
+  return <React.Fragment>
+    {loadNextSubject && <Navigate to={getNextAuditPath(auditSubject)}/>}
+    <TextInputWithLabel
+      containerStyle={{display: 'inline-block'}}
+      style={{width: '15rem', margin: '1rem'}}
+      labelText = {buttonLabel || auditSubjectType}
+      value = {auditSubject}
+      onChange = {setAuditSubject}
+    />
+    <Button style={{height: '1.5rem', margin: '0.25rem 0.5rem'}}
+            disabled={fp.isEmpty(auditSubject)}
+            onClick={() => setLoadNextSubject(true)}>
+    Audit
+    </Button>
+  </React.Fragment>;
+};
+
+const NumActions = ({onChange, totalActions}) => {
+  const [displayNum, setDisplayNum] = useState(20);
+
+  useEffect(() => {
```

Comment:
I am going to add a comment here to make it clear what the purpose of this is: 
To debounce the input action. Without this it will re-render immediately after every number change.

Reply:
I added a useDebounce hook rather than a code comment. Hopefully this will make it clearer in the code what we are doing rather than using a comment to explain.

Reply:
Yeah, I remember it from a EE course. It does seem to help.